module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,44 @@
+#include "compression.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <utility>
+
+vector_of_pairs compressGrayscale(const array_of_arrays& bitmap) {
+    vector_of_pairs compressed{};
+    for (const auto& row : bitmap) {
+        for (auto it = row.begin(); it != row.end();) {
+            auto findIt = std::adjacent_find(it, row.end(), std::not_equal_to<int>{});
+            if (findIt != row.end()) {
+                findIt++;
+            }
+            compressed.emplace_back(*it, std::distance(it, findIt));
+            it = findIt;
+        }
+    }
+    return compressed;
+}
+
+array_of_arrays decompressGrayscale(const vector_of_pairs& bitmap) {
+    array_of_arrays decompressed{};
+    auto it = decompressed.front().begin();
+    std::for_each(bitmap.begin(), bitmap.end(), [it](const auto& pair) mutable {
+        it = std::fill_n(it, pair.second, pair.first);
+    });
```

Comment:
Can you please explain me, how does it fill all the rows of the output map? I see that "it" is an iterator that is modified after inputting groups of pixels, but how does this "it" "jump" to the next row to fill it? Am I thinking right, that you're using the fact, that in an "array of arrays" consecutive arrays are also adjacent in memory, and when iterator exceeds end of array in given row it simultanously starts to iterate through next row? (just like one would use incrementing a pointer on a raw two-dimensional array)